### ISSUE TITLE
AssemblyScript Public Meeting #24 - November 4th, 2020

### ISSUE THREAD

**Issue by torch2424** [AssemblyScript/working-group#49]
# Date and Time

Per the last meeting #48 , Daylight savings in the US went into affect, so please see the updated time :smile: (Non-US shouldn't have to worry I don't think, the UTC time stays the same).

This public meeting will take place: October 21st, 2020, 18:00 UTC (**10:00 AM US PDT, UTC -8**)

# General Agenda

* Agenda Items from comments left on this Github issue
* Additional in-meeting comments / discussion
* [If time allows] Recap of [WebAssembly CG meeting](https://github.com/WebAssembly/meetings) if anyone attended

Feel free to comment on this issue if you have any agenda items you would like to bring up. 
Meeting Notes will be placed on this issue for those who cannot make the meeting.

# Meeting Information

We will use the AssemblyScript discord for meetings: https://discord.gg/vSBbJ5T . We will meet in the "meeting" voice channel :+1: 

Anyone who is contributing to the AssemblyScript project, building something with AssemblyScript, interested in the Assembly project or WebAssembly in general, is welcome to join! 😄

Notes will be taken by the host, and posted after the meeting. Notes are free to be edited through comments on the meeting notes at a later time.

**Comment by torch2424** [comment]
* @sirwillem General update on https://github.com/nearprotocol/assemblyscript-json . Fastly would like to help and contribute! :) 

* Would it be a good idea to add a feature to the compiler to embed static files? We have `memory.data`, but `import`ing a static file would be cool :smile:

**Comment by torch2424** [comment]
# Meeting Notes

* AS JSON
   * https://github.com/nearprotocol/assemblyscript-json 
   * @willemneal not here, will ping on the discord
* Static file importing
   * E.g, a rollup or webpack static file import syntax? Should it be part of the compiler? Or another tool?
   * https://github.com/AssemblyScript/assemblyscript/pull/1370
   * @torch2424 I might build a quick tool for Fastly, and then move into the compiler.
* @dcodeIO : General Fastly AS Experience? Anything AS is doing wrong?
   * @torch2424 : Only thing I heard was closures. Most other complaints are per Fastly specific stuff (Our platform, or small things I missed in the SDK).
      * I think we are doing really solid, and the general AS experience is awesome :)
      * Also! Someone on Fastly Devrel had to independently learn AS, and then write the intro docs for new users. They picked it up super fast. They found some rough edges, but for the most part, they really enjoyed it and wasn’t too bad to pick up :)
* @dcodeIO : Runtime presentation
   * We’re in a dilemma. 
      * Wasm GC spec isn’t really working out in our favor. 
      * Trying to improve our current one isn’t really making too much progress either.
      * Tracing GC is not a good fit
   * GC within Wasm MVP has challenged
      * No random access to the execution stack. 
      * Moving and copying has consequences.
         * Since we have an external host, pointers and things can be moved. Thus, we need to be informed.
      * Generational GC is not a good fit.
   * What about reference counting?
      * That is what we currently do.
      * If something has at least one reference, than it will be kept alive.
         * Functions: Must maintain RC+1 for locals
         * Blocks: Must maintain RC+1 for return or temporary values
      * Problems:
         * Frequent reference count increment and decrement. Which leads to write barriers, which leads to GC pressure
         * Complexity that we can’t offload to a separate pass.
            * This adds a TON of complexity to code gen. Which makes this really hard to continue using
            * ASC is mostly single pass, reusing Binaryen IR
            * Binaryen cannot reason about higher-level managed types.
            * Hard to maintain, hard to debug, -> stale progress.
      * Typically not generational
   * RC: Autorelease pools?
      * Idea: Put the values with RC+1 into a pool to batch release later. For example, when exiting a block or function, just release everything
         * Swift is currently doing what we are doing, but they have a whole compiler team that can handle this
         * This would be more granular, it’s less efficient with more shorter pauses, and smaller memory footprint
         * Or could be more granular, more efficient with less longer pauses, and a larger memory footprint
      * Trades compiler complexity for runtime overhead
         * Tracking state during codegen, maintain pools at runtime
            * Still requires some compiler cooperation, but way simpler
         * Release statically when local goes out if scope -> batch release a pool
         * Maintaining RC+1 for locals -> Fire and forget autorelease, a bit closer to tracing.
   * RC: Efficient autorelease pools?
      * Should avoid allocating pools dynamically
         * Linked lists (one more header word at our disposal)
      * Should avoid duplicate items in pools
         * Hierachical, mutually exclusive lists
            * RC+1, not RC+X
      * Still not great :(
         * Defers RC work, but doesn’t reduce it
         * Unreclaimed memory overhead. E.g in loops, in between pool releases.
         * For example:
            * If we have a loop of a million iterations
            * We can allocate and release on ever iteration
               * Slow, but uses less memory
            * Or release once we leave the loop
               * Fast, but allocates like a million objects
   * @dcodeIO : A little worried we won’t be able to get the runtime to a state that our current users would want. Since we don’t have a whole dedicated compiler team where we could write our entire own IR, and Wasm GC is having a hard time making progress.
      * @torch2424 : I think we are doing fine :) Personally, I’d say we go for the less efficient runtime option. As the runtime is removable. And I think the bigger complaints right now are with ease of use, rather than with runtime performance. Especially since we are mostly just running up against the limits of the MVP not much we can do.
      * @torch2424 : Though, on second thought, I do understand your worry when you think of the micro benchmark articles on performance. But again. I think my points above still stand.
   * @MaxGraey : Might be good to just do what we can now on top of the current MVP. Get it in the compiler. And then add custom IR, or build on top of new specs later :) 
* 〖ƝƁƘ〗✧Ṃ!ṨҬ!Ḱ✧ (Discord Name): Using Assembly script in my game
   * Great for handling bots. Was able to pick up Wasm super easy using AS, and didn’t have to use another language.
* 〖ƝƁƘ〗✧Ṃ!ṨҬ!Ḱ✧ (Discord Name): Obfuscation?
   * @dcodeIO : they want to rename function exports etc. to something unrecognizable or empty, but stripping the debug section cannot do that (for exports, only for non-exports)

---

Then we got raided by a discord group :joy: Action Item: Figure out how to avoid random bots and people joining the meeting randomly :upside_down_face:

**Comment by torch2424** [comment]
Closing this in favor of the next meeting: https://github.com/AssemblyScript/working-group/issues/50

Feel free to continue discussions from this meeting here! :smile:
